File: stealth-browser-lab/src/humanMouse.js
'use strict';
/**
 * 人間らしいマウス軌跡の生成。
 *
 * 中核は「WindMouse」アルゴリズム。
 * 目標地点への「重力(gravity)」と、ランダムに揺れる「風(wind)」を
 * 速度ベクトルに加え続けることで、直線ではない自然な弧・微妙な揺れ・
 * 終端での減速を再現する。ボット検出は「点A→点Bへ完全な直線で一瞬移動」
 * のような非人間的な軌跡を疑うため、ここが効く。
 *
 * 参考: BenLand100 による WindMouse（公開アルゴリズム）。
 */
const { sleep, gaussianClamped, rand } = require('./utils');

const SQRT3 = Math.sqrt(3);
const SQRT5 = Math.sqrt(5);

/**
 * 始点(x0,y0)→終点(x1,y1) の軌跡座標列を返す。
 * @param {object} o 調整パラメータ
 *   G: 重力（大きいほど真っ直ぐ目標へ）
 *   W: 風（大きいほど揺れる）
 *   maxStep: 1ステップの最大移動量（大きいほど速い）
 *   targetArea: 終端で減速し始める距離
 * @returns {Array<[number,number]>}
 */
function windMouse(x0, y0, x1, y1, o = {}) {
  const G = o.G ?? 9;
  const W = o.W ?? 3;
  let maxStep = o.maxStep ?? 15;
  const targetArea = o.targetArea ?? 12;

  const points = [];
  let vx = 0;
  let vy = 0;
  let wx = 0;
  let wy = 0;
  let x = x0;
  let y = y0;

  let dist = Math.hypot(x1 - x, y1 - y);
  // 無限ループ保護
  let guard = 0;
  while (dist >= 1 && guard++ < 10000) {
    const wMag = Math.min(W, dist);
    if (dist >= targetArea) {
      // 目標から遠い間は風がランダムに揺らす
      wx = wx / SQRT3 + ((2 * Math.random() - 1) * wMag) / SQRT5;
      wy = wy / SQRT3 + ((2 * Math.random() - 1) * wMag) / SQRT5;
    } else {
      // 目標に近づいたら風を減衰させ、ステップも細かくして「吸い付く」
      wx /= SQRT3;
      wy /= SQRT3;
      if (maxStep < 3) maxStep = rand(3, 6);
      else maxStep /= SQRT5;
    }
    // 風 + 重力（目標方向）を速度に加える
    vx += wx + (G * (x1 - x)) / dist;
    vy += wy + (G * (y1 - y)) / dist;

    // 速度が maxStep を超えたらクリップ（=最高速の制限）
    const vMag = Math.hypot(vx, vy);
    if (vMag > maxStep) {
      const vClip = maxStep / 2 + rand(0, maxStep / 2);
      vx = (vx / vMag) * vClip;
      vy = (vy / vMag) * vClip;
    }

    x += vx;
    y += vy;
    points.push([Math.round(x), Math.round(y)]);
    dist = Math.hypot(x1 - x, y1 - y);
  }
  points.push([Math.round(x1), Math.round(y1)]);
  return points;
}

/**
 * Playwright の page でマウスを人間らしく移動させる。
 * 各点の間に微小なゆらぎ待機を入れて、移動速度に「ムラ」を作る。
 */
async function moveMouseHuman(page, toX, toY, opts = {}) {
  const from = opts.from || page.__lastMouse || { x: rand(0, 200), y: rand(0, 200) };
  const path = windMouse(from.x, from.y, toX, toY, opts);
  for (const [px, py] of path) {
    await page.mouse.move(px, py);
    // 1点ごとの待機 (ms)。短いが一定でないのがポイント。
    await sleep(gaussianClamped(7, 4, 1, 25));
  }
  page.__lastMouse = { x: toX, y: toY };
}

/**
 * 要素まで人間らしく動いてクリック。
 * 要素中心ぴったりではなく、少しズラした点を狙うのも非ボットらしさ。
 */
async function clickHuman(page, selector, opts = {}) {
  const el = await page.waitForSelector(selector, { state: 'visible' });
  const box = await el.boundingBox();
  if (!box) throw new Error(`要素の座標が取れません: ${selector}`);
  const targetX = box.x + box.width / 2 + rand(-box.width * 0.2, box.width * 0.2);
  const targetY = box.y + box.height / 2 + rand(-box.height * 0.2, box.height * 0.2);
  await moveMouseHuman(page, targetX, targetY, opts);
  await sleep(gaussianClamped(90, 40, 30, 250)); // 押下前の「ためらい」
  await page.mouse.down();
  await sleep(gaussianClamped(60, 25, 20, 160)); // 押している時間
  await page.mouse.up();
}

module.exports = { windMouse, moveMouseHuman, clickHuman };

File: stealth-browser-lab/stealth_lab/human_mouse.py
"""人間らしいマウス軌跡の生成（WindMouse アルゴリズム）。

目標への「重力(gravity)」とランダムに揺れる「風(wind)」を速度ベクトルに
加え続けることで、直線でない自然な弧・微妙な揺れ・終端の減速を再現する。
検出側は「点A→点Bへ完全な直線で一瞬移動」を疑うため、ここが効く。

参考: BenLand100 による WindMouse（公開アルゴリズム）。
"""
import math
import random

from .utils import rand, gaussian_clamped, sleep_ms

SQRT3 = math.sqrt(3)
SQRT5 = math.sqrt(5)

# ページごとの最後のマウス座標を覚えておく（id(page) -> (x, y)）
_last_pos: dict[int, tuple[float, float]] = {}


def wind_mouse(x0, y0, x1, y1, G=9.0, W=3.0, max_step=15.0, target_area=12.0):
    """始点(x0,y0)→終点(x1,y1) の軌跡座標列 [(x, y), ...] を返す。

    G: 重力（大きいほど真っ直ぐ目標へ）
    W: 風（大きいほど揺れる）
    max_step: 1ステップの最大移動量（大きいほど速い）
    target_area: 終端で減速し始める距離
    """
    points = []
    vx = vy = wx = wy = 0.0
    x, y = float(x0), float(y0)
    dist = math.hypot(x1 - x, y1 - y)
    guard = 0
    while dist >= 1 and guard < 10000:
        guard += 1
        w_mag = min(W, dist)
        if dist >= target_area:
            # 目標から遠い間は風がランダムに揺らす
            wx = wx / SQRT3 + (2 * random.random() - 1) * w_mag / SQRT5
            wy = wy / SQRT3 + (2 * random.random() - 1) * w_mag / SQRT5
        else:
            # 目標に近づいたら風を減衰、ステップも細かくして「吸い付く」
            wx /= SQRT3
            wy /= SQRT3
            if max_step < 3:
                max_step = rand(3, 6)
            else:
                max_step /= SQRT5
        # 風 + 重力（目標方向）を速度に加える
        vx += wx + G * (x1 - x) / dist
        vy += wy + G * (y1 - y) / dist
        v_mag = math.hypot(vx, vy)
        if v_mag > max_step:
            v_clip = max_step / 2 + rand(0, max_step / 2)
            vx = vx / v_mag * v_clip
            vy = vy / v_mag * v_clip
        x += vx
        y += vy
        points.append((round(x), round(y)))
        dist = math.hypot(x1 - x, y1 - y)
    points.append((round(x1), round(y1)))
    return points


async def move_mouse_human(page, to_x, to_y, frm=None):
    """page のマウスを人間らしく移動。各点間に微小なゆらぎ待機を入れる。"""
    start = frm or _last_pos.get(id(page)) or (rand(0, 200), rand(0, 200))
    for px, py in wind_mouse(start[0], start[1], to_x, to_y):
        await page.mouse.move(px, py)
        await sleep_ms(gaussian_clamped(7, 4, 1, 25))
    _last_pos[id(page)] = (to_x, to_y)


async def click_human(page, selector, **kw):
    """要素まで人間らしく動いてクリック。中心ぴったりでなく少しズラす。"""
    el = await page.wait_for_selector(selector, state="visible")
    box = await el.bounding_box()
    if not box:
        raise RuntimeError(f"要素の座標が取れません: {selector}")
    tx = box["x"] + box["width"] / 2 + rand(-box["width"] * 0.2, box["width"] * 0.2)
    ty = box["y"] + box["height"] / 2 + rand(-box["height"] * 0.2, box["height"] * 0.2)
    await move_mouse_human(page, tx, ty)
    await sleep_ms(gaussian_clamped(90, 40, 30, 250))  # 押下前の「ためらい」
    await page.mouse.down()
    await sleep_ms(gaussian_clamped(60, 25, 20, 160))   # 押している時間
    await page.mouse.up()
